lib/SILGen/SILGenLValue.cpp

```diff
@@ -1344,6 +1351,86 @@ namespace {
       assert(getAccessorDecl()->isSetter());
       SILDeclRef setter = Accessor;
 
+      if (hasPropertyDelegate() && IsOnSelfParameter) {
+        // This is delegated property. Instead of emitting a setter, emit an
+        // assign_by_delegate with the allocating initializer funcction and the
+        // setter function as arguments. DefiniteInitializtion will then decide
+        // between the two functions, depending if it's an initialization or a
+        // re-assignment.
+        //
+        VarDecl *field = dyn_cast<VarDecl>(Storage);
+        VarDecl *backingVar = field->getPropertyDelegateBackingVar();
+        assert(backingVar);
+        CanType ValType = backingVar->getType()->getCanonicalType();
```

Comment:
If the initializer is itself generic this will be written in terms of the wrong archetypes. I think you should use the interface type here and substitute the right substitutions.

Reply:
There are restrictions on how the initializer can look like.
To get the substitutions, I basically did the same as Doug did in SILGenConstructor.cpp: emitImplicitValueConstructor (https://github.com/DougGregor/swift/commit/e674377b44aa92b58114103eb2154aa317414ee2#diff-0d8359a6c47d13683f989e2f711454c5R174)

Reply:
I think both copies of the code are not going to work when the initializer has different archetypes from the type. That’s why I wanted you to try an init with it’s own generic parameters.

Good point about the different substitutions though. It’s worth an explanatory comment.

Reply:
FWIW, the form of `init(initialValue:)` is locked down to have the same generic signature as the type. I am (separately) turning this into an expression so the type checker forms all of the calls for us.